### ISSUE TITLE
data flow: possible values not constrained by control flow (indirect)

### ISSUE THREAD

**Issue by pmackinlay** [Vector35/binaryninja-api#968]
Binary Ninja Version: 1.1.1068-dev Personal, beb4a48f
Platform: Windows 10 (10.0)

This is a very similar issue to #953, with the difference being that the flow condition is based on a copy of the value in question.

Referring to the screenshot below, BN correctly determines the range of values for r0 @ 6 is -0x8000..0x7fff, but fails to further constrain that range through the control flow @ 8, although it should be able to unambiguously do so due to the expression @ 7.

In this case, BN also fails to create the outward edges from @ 16, probably because the unconstrained range is too large.

![image](https://user-images.githubusercontent.com/352356/36772805-3ce237ce-1c89-11e8-8737-e495d2ff9c78.png)

**Comment by joshwatson** [comment]
This is probably related to #801

**Comment by bpotchik** [comment]
I don't think this will be fixed by #801. We will have to add an algebraic solver for this pattern. I will look into this at some point.

**Comment by bpotchik** [comment]
Just an update. I have a worker solver for this and will push pending some validation and performance tests.
![screen shot 2018-03-12 at 11 20 28 am](https://user-images.githubusercontent.com/29309548/38045455-cf140732-328a-11e8-9c73-8c4be580fa69.png)

**Comment by bpotchik** [comment]
Fixed in `5.3.9237-dev`.
